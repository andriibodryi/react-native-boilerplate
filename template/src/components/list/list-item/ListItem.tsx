import * as React from 'react';
import { Pressable, View } from 'react-native';

import { Text } from '@src/components/typography';
import { Icon } from '@src/components';

import { stylePresets } from './list-item.presets';
import { ListItemTypes } from './list-item.types';

const ListItem: React.FC<ListItemTypes> = ({
  label,
  sublabel,
  style,
  labelStyle,
  sublabelStyle,
  preset = 'basic',
  sublabelNumberOfLines = 1,
  labelNumberOfLines = 1,
  headingColor = 'secondary',
  rightIconColor,
  rightIcon,
  leftIcon,
  leftIconColor,
  backgroundColor,
  ...rest
}) => {
  const styles = stylePresets[preset];

  return (
    <Pressable {...rest} style={[styles.container, style]}>
      <View style={styles.row}>
        {leftIcon && (
          <Icon
            size={24}
            name={leftIcon}
            color={leftIconColor}
            containerStyle={[
              styles.actionIconContainer,
              styles.marginVerticalNone,
              { backgroundColor },
            ]}
          />
        )}
        <View style={[styles.item, styles.content]}>
          <Text
            color={headingColor}
            preset="textSmMedium"
            selectable={false}
            style={labelStyle}
            numberOfLines={labelNumberOfLines}
          >
            {label}
          </Text>
          {Boolean(sublabel) && (
            <Text
              preset="textSmRegular"
              selectable={false}
              style={sublabelStyle}
              numberOfLines={sublabelNumberOfLines}
            >
              {sublabel}
            </Text>
          )}
        </View>

        {rightIcon && (
          <Icon
            size={24}
            name={rightIcon}
            color={rightIconColor}
            containerStyle={[
              styles.actionIconContainer,
              styles.iconMarginRight,
              styles.marginVerticalNone,
              { backgroundColor },
            ]}
          />
        )}
      </View>
    </Pressable>
  );
};

export default ListItem;
