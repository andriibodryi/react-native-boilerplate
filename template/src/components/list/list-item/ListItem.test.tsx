import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';

import ListItem from './ListItem';

describe('ListItem', () => {
  it('should render list item correctly', () => {
    const tree = renderer.create(<ListItem label="list item title" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should contain "double tap to activate" accessibility hint by default', () => {
    const { queryByA11yHint } = render(<ListItem label="list item title" />);

    const payAContactButton = queryByA11yHint(/double tap to activate/i);

    expect(payAContactButton).toBeTruthy();
  });

  it('should not contain "double tap to activate" accessibility hint for descriptive list preset', () => {
    const { queryByA11yHint } = render(
      <ListItem label="list item title" preset="descriptiveList" />,
    );

    const payAContactButton = queryByA11yHint(/double tap to activate/i);

    expect(payAContactButton).toBeFalsy();
  });
});
