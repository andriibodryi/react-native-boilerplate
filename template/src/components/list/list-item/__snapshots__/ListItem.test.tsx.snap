// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should render list item correctly 1`] = `
<View
  accessible={false}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
        "width": "100%",
      },
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ],
    ]
  }
  testID="list_item"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginVertical": 8,
            "paddingLeft": 8,
          },
          Object {
            "flex": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        accessible={false}
      >
        <Text
          accessibilityHint="double tap to activate"
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#4D4D4D",
                "fontFamily": "Aspira-Regular",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 20,
              },
              Array [
                Object {
                  "fontFamily": "Aspira-Medium",
                  "fontSize": 16,
                  "fontWeight": "500",
                  "lineHeight": 20,
                },
                Object {
                  "textAlign": "left",
                },
                Object {
                  "color": "rgb(77, 77, 77)",
                },
                undefined,
              ],
            ]
          }
          testID="list_item_label"
        >
          list item title
        </Text>
      </View>
    </View>
  </View>
</View>
`;
